### PR TITLE
Unify and correct archive states

### PR DESCRIPTION
**Pull request by theefer**
Rework archive buttons on search and image page to be more consistent throughout in the use of icons, actions available (if auto-persisted, just show disabled control with "Kept in Library"), etc.

Also move some logic into a new logic service that is more flexible than the current one-size-fits-all image-service.

recorded

Once this is out, we can review what icon would be best to represent the three states of auto-persisted, manually archived or not archived at all.

### CODE REVIEW

File: kahuna/public/js/components/gr-archiver/gr-archiver.html [hunk 1]
```diff
@@ -1,11 +1,20 @@
 <div class="batch-archive"
      ng:switch="ctrl.archivedState">
+    <span ng:switch-when="kept"
+          type="button"
```
Comment (akash1810) [inline kahuna/public/js/components/gr-archiver/gr-archiver.html]: is `type` a valid attribute for a `span`?

Comment (theefer) [inline kahuna/public/js/components/gr-archiver/gr-archiver.html]: Nope. Thanks, will fix!

File: kahuna/public/js/components/gr-archiver/gr-archiver.js [hunk 2]
```diff
@@ -58,6 +67,21 @@ module.controller('grArchiverCtrl', [
...
+        function listAllPersistenceExplanations(images) {
+            return images.
+                map(imageLogic.getPersistenceExplanation).
+                map(items => new Set(items)).
+                reduce((all, items) => all.union(items)).
+                toArray();
+        }
+
+        // Takes an array and generates a '1, 2, 3 or 4' string
+        function humanUnion(list) {
```
Comment (akash1810) [inline kahuna/public/js/components/gr-archiver/gr-archiver.js]: nice

File: kahuna/public/js/services/image-logic.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Helpers to apply business logic on image resources.
+ */
+imageLogic.factory('imageLogic', ['imageAccessor', function(imageAccessor) {
+
+    function canBeDeleted(image) {
+        return image.getAction('delete').then(action => !! action);
+    }
+
+    function canBeArchived(image) {
```
Comment (akash1810) [inline kahuna/public/js/services/image-logic.js]: minor: is it worth calling this `canBePersisted` to have a common language? (ideally, would be done everywhere though :neutral_face: )

**Comment by akash1810** [comment]
:+1:

**Comment by kenoir** [comment]
@theefer how does this look with the meta panel locked? Does it result in two lock icons showing next to each other with different functions?

File: kahuna/public/js/services/image-logic.js (hunk 3 above)
Comment (theefer) [inline kahuna/public/js/services/image-logic.js]: I thought there was a conscious use of both words to mean different things: archived = user toggle, persisted = derived state of whether the image can be deleted, with archived = true implying persisted = true?

Comment (akash1810) [inline kahuna/public/js/services/image-logic.js]: ah, yep :blush:

**Comment by theefer** [comment]
@kenoir It's basically crap:

image

Although it's worst form is when the screen is too narrow and we fallback to icons only:

image

My plan is to replace the icons used for archive state with something else than a lock.

**Comment by theefer** [comment]
Note that I've already moved the archive button into the action bar in a previous commit so this is already the case.

**Pull request merged**